README.rst
Picomongo
#########

Faster than minimongo, lighter than micromongo, please welcome picomongo, the ultimate Mongo ODM made by Dailymotion.

Getting started
===============

To start off with picomongo, just import it::

    >>> from picomongo import Document, ConnectionManager
    >>> ConnectionManager.configure()

And you're ready, let's define a document::

    >>> class UserDocument(Document):
    ...     pass
    >>> user = UserDocument({'name': 'Mike'})
    >>> user
    UserDocument({'name': 'Mike'})
    >>> user.name
    'Mike'
    >>> user.save()

YOU: Wait, wait where is my document ?

Don't worry, in order to facilitate your work, picomongo use these default values:

* Mongo uri: 'mongodb://localhost'
* Database: 'test'
* Collection: Your document class name in lowercase

You have access to these values, which are traditionnal pymongo objects::

    >>> UserDocument.con
    Connection('localhost', 27017)
    >>> UserDocument.db
    Database(Connection('localhost', 27017), u'test')
    >>> UserDocument.col
    Collection(Database(Connection('localhost', 27017), u'test'), u'userdocument')


One thing you should always keep in mind, you always need to call configure before be able to save/retrieve your documents. If you want to use the default configuration, just call configure without arguments otherwise see below (part Configuration time).

TIP: You can call configure after the declaration of your documents BUT before document saving/retrieving.

You can use them as you will do with traditionnal pymongo objects, for example you can retrieve your user using traditionnal collection::

    >>> UserDocument.col.find_one()
    {u'_id': ObjectId('4eb2cae58250f05eb4000000'), u'name': u'Mike'}

YOU: But wait, why I get a dict, I want an object instead.

Don't worry, it's even more simpler::

    >>> user2 = UserDocument.find_one()
    UserDocument({u'_id': ObjectId('4eb2cae58250f05eb4000000'), u'name': u'Mike'})
    >>> user2.name
    u'Mike'

But when this auto-configuration was done? As soon as you try to access them.

Configure the collection
========================

You can configure on which collection your documents will be saved. You can override the collection_name in your custom Document classes if you want to configure it. Example::

    >>> class CustomDocument(Document):
    ...     collection_name = 'my_custom_collection'
    >>> custom = CustomDocument()
    >>> custom.col
    Collection(Database(Connection('localhost', 27017), u'test'), u'my_custom_collection')

Configuration time
==================

Once you use it a bit, let's see the most powerful part of picomongo, configuration.

All configurations are stored in ConnectionManager::

   >>> from picomongo import Document, ConnectionManager

You can add your own configuration by calling the configure method with your configuration. The configuration format is::

    {'_default_': {'uri': 'default_uri', 'db': 'default_db'},
     'document_name': {'uri': 'specific_uri', 'db': 'specific_db', 'col': 'default_col'},
     'document_name2': ...,}

Uri must be a valid mongodb connection uri as described in this doc page: http://www.mongodb.org/display/DOCS/Connections

Nothing is required in the configuration, and picomongo will use some rules to compute the final configuration:

* In default configuration:
    * If uri is not present, use 'mongodb://localhost'
    * If db is not present, use 'test'
* In document configuration:
    * If uri is not present, use default uri
    * If db is not present, use default db
    * If col is not preset, use document class name

You can access configuration using this syntax::

    >>> ConnectionManager.get_config('_default_')   # Access default configuration
    >>> ConnectionManager.get_config('document')    # Access configuration for document named 'document'

Here is some examples of configurations:

* Change default db::

    >>> ConnectionManager.configure({'\_default\_': {'db': 'other_db'}})
    >>> ConnectionManager.get_config('\_default\_').db
    Database(Connection('localhost', 27017), u'other_db')

* Store some documents in another mongodb instance::

    >>> ConnectionManager.configure({'document1': {'uri': 'mongodb://127.0.0.1:8000'}})
    >>> ConnectionManager.get_config('_default_').con
    Connection('localhost', 27017)
    >>> ConnectionManager.get_config('document1').con
    Connection('localhost', 8000)

TIP: This last example will surely fail as picomongo try to connect to this uri during configuration (and you probably do not have a mongodb instance running at this uri).


debian/changelog
picomongo (0.8-1~precise1) precise-dc; urgency=low

  * Read preference to PRIMARY_PREFERED

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Sat, 17 Aug 2013 15:23:30 +0200

picomongo (0.7-1~precise1) precise-dc; urgency=low

  * ensure_index background parameter is True by default

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Thu, 28 Feb 2013 15:22:17 +0100

picomongo (0.6-1~precise1) precise-dc; urgency=low

  * Precise

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Sun, 09 Sep 2012 14:17:49 +0200

picomongo (0.6-1~karmic1) karmic-dc; urgency=low

  * Add support for ReplicaSet

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Mon, 25 Jun 2012 10:51:08 +0200

picomongo (0.5-1~karmic1) karmic-dc; urgency=low

  * Bug fix

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Mon, 13 Feb 2012 17:53:49 +0100

picomongo (0.4-1~karmic1) karmic-dc; urgency=low

  * Bug fix

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Tue, 31 Jan 2012 11:42:30 +0100

picomongo (0.3-1~karmic1) karmic-dc; urgency=low

  * Bug fix

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Fri, 13 Jan 2012 11:46:11 +0100

picomongo (0.2-1~karmic1) karmic-dc; urgency=low

  * New release

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Thu, 12 Jan 2012 17:33:38 +0100

picomongo (0.1-1~karmic1) karmic-dc; urgency=low

  * Initial release

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Wed, 16 Nov 2011 16:33:57 +0100


tests/utils.py
def Call(*args, **kwargs):
    return (args, kwargs)


tests/test_document.py
import unittest

import pymongo
from mock import patch, Mock, sentinel

from bson.objectid import ObjectId

from pymongo import Connection
from pymongo.database import Database
from pymongo.collection import Collection
from pymongo.errors import InvalidOperation, DuplicateKeyError, OperationFailure

from picomongo import Document, ConnectionManager
from picomongo.exceptions import ValidationError
from utils import Call

#Examples document class
class UserDocument(Document):
    def full_name(self):
        return '%s %s' % (self.first_name, self.name)

class SimpleIndex(Document):
    indexes = []

class DefaultDocument(Document):
    default_values = {'a': 1, 'b': 2}

class ValidationDocument(Document):
    pass

class DocumentBaseTestCase(unittest.TestCase):

    def test_init_with_values(self):
        data = {'attr1': 1, 'attr2': 2}
        document = Document(data)

        self.assertEquals(document, data)

    def test_init_with_private_names(self):
        data = {'_private': 42}
        self.assertRaises(ValueError, Document, data)

    def test_access_config(self):
        ConnectionManager.configure()

        self.assertTrue(isinstance(UserDocument.col, Collection))
        self.assertTrue(isinstance(UserDocument.db, Database))
        self.assertTrue(isinstance(UserDocument.con, Connection))

    def test_reconfiguration(self):
        ConnectionManager.configure()

        self.assertEqual(UserDocument.db.name, 'test')

        ConnectionManager.configure({'userdocument': {'db': 'test2'}})
        self.assertEqual(UserDocument.db.name, 'test2')

class DocumentPersistenceBaseTestCase(unittest.TestCase):

    def setUp(self):
        self.document = Document()

        self.connection = pymongo.Connection()
        self.db = self.connection.test
        self.collection = self.db.test_collection

    def tearDown(self):
        self.collection.remove()

    def test_insert(self):
        """Test that a document can be directly insert into mongodb.
        """
        self.document.attribute = 'value'

        inserted_id = self.collection.insert(self.document)

        self.assertTrue(isinstance(self.document._id, ObjectId))

        self.assertEqual(inserted_id, self.document._id)
        self.assertEqual(self.collection.find_one(), {'attribute': 'value', '_id': self.document._id})

    def test_document_rehydration(self):
        """Test that mongodb document can be rehydrated into Document classes.
        """
        user_data = {'first_name': 'Boris', 'name': 'FELD'}

        self.collection.insert(user_data)

        user = self.collection.find_one(as_class=UserDocument)

        self.assertEqual(user.full_name(), 'Boris FELD')

class DocumentPersistenceTestCase(unittest.TestCase):

    def setUp(self):
        self.data = {'first_name': 'Boris', 'name': 'FELD'}
        self.document = UserDocument()
        self.document.first_name = 'Boris'
        self.document.name = 'FELD'

    def tearDown(self):
        self.document.col.remove()

    def test_save(self):
        self.document.save()

        self.assertEqual(UserDocument.col.find_one(), self.document)

    def test_rehydrate_from_class(self):
        self.document.save()

        document = UserDocument.find_one()

        self.assertEqual(document.full_name(), 'Boris FELD')

class DocumentDBMethodsTestCase(unittest.TestCase):

    def tearDown(self):
        UserDocument.col.remove()

    def test_find_one(self):
        args = ['arg1', 'arg2', 'arg3']
        kwargs = {'kwarg1': '', 'kwarg2': '', 'kwarg3': ''}
        with patch.object(UserDocument, 'col') as mock_col:
            mock_find_one = mock_col.find_one
            mock_find_one.return_value = {}
            UserDocument.find_one(*args, **kwargs)

        self.assertEqual(mock_find_one.call_args_list, [Call(*args, **kwargs)])

    def test_fine_one_no_document(self):
        self.assertEqual(UserDocument.find_one(), None)

    def test_find(self):
        args = ['arg1', 'arg2', 'arg3']
        kwargs = {'kwarg1': '', 'kwarg2': '', 'kwarg3': ''}
        with patch.object(UserDocument, 'col') as mock_col:
            mock_find = mock_col.find
            UserDocument.find(*args, **kwargs)

        self.assertEqual(mock_find.call_args_list, [Call(*args, **kwargs)])

    def test_find_no_document(self):
        self.assertEqual(list(UserDocument.find()), [])

    def test_delete(self):
        user = UserDocument()
        user.save()

        args = ['arg1', 'arg2', 'arg3']
        kwargs = {'kwarg1': '', 'kwarg2': '', 'kwarg3': ''}

        with patch.object(UserDocument, 'col') as mock_col:
            mock_remove = mock_col.remove
            user.delete(*args, **kwargs)

        args.insert(0, {'_id': user._id})
        self.assertEqual(mock_remove.call_args_list, [Call(*args, **kwargs)])

    def test_remove_unsaved(self):
        user = UserDocument()

        self.assertRaises(InvalidOperation, user.delete)

    def test_save(self):
        user = UserDocument()

        kwargs = {'kwarg1': '', 'kwarg2': '', 'kwarg3': ''}

        with patch.object(UserDocument, 'col') as mock_col:
            mock_save = mock_col.save
            user.save(reload=False, **kwargs)

        args = [user]
        self.assertEqual(mock_save.call_args_list, [Call(*args, **kwargs)])

    def test_save_private_attr(self):
        document = UserDocument()
        document.first_name = 'Boris'
        document.name = 'FELD'
        document.save()

        self.assertEqual(UserDocument.col.find_one(), {'first_name': 'Boris',
            'name': 'FELD', '_id': document._id})

    def test_reload(self):
        document = UserDocument()
        document.first_name = 'Boris'
        document.name = 'FELD'
        document.save()

        UserDocument.col.update({'_id': document._id},
                                {'$set': {'name': 'SMITH'}})

        document.reload()

        self.assertEqual(document,
            {'_id': document._id, 'first_name': 'Boris', 'name': 'SMITH'})

    def test_reload_error(self):
        user = UserDocument()

        self.assertRaises(InvalidOperation, user.reload)

    def test_reload_invalid_id(self):
        user = UserDocument()
        user['_id'] = ObjectId('1'*24)

        self.assertRaises(OperationFailure, user.reload)

    def test_reload_remove_fields(self):
        document = UserDocument()
        document.first_name = 'Boris'
        document.name = 'FELD'
        document.age = 21
        document.save()

        UserDocument.col.update({'_id': document._id},
                                {'$unset': {'age': 1}})
        document.reload()

        self.assertEqual(document,
            {'_id': document._id, 'first_name': 'Boris', 'name': 'FELD'})

    def test_reload_remove_fiels_and_private_fiels(self):
        document = UserDocument()
        document.first_name = 'Boris'
        document.name = 'FELD'
        document.age = 21
        document._private = 42
        document.save()

        self.assertEqual(UserDocument.col.find_one(), {'first_name': 'Boris',
            'name': 'FELD', 'age': 21, '_id': document._id})
        UserDocument.col.update({'_id': document._id},
                                {'$unset': {'age': 1}})
        document.reload()

        self.assertEqual(document,
            {'_id': document._id, 'first_name': 'Boris', 'name': 'FELD'})
        self.assertEqual(document._private, 42)

    def test_advanced_find(self):
        UserDocument.col.insert({'first_name': 'Boris', 'name': 'FELD'}, safe=True)
        UserDocument.col.insert({'first_name': 'John', 'name': 'SMITH'}, safe=True)
        UserDocument.col.insert({'first_name': 'Mickael', 'name': 'JOHNSON'}, safe=True)

        expected_values = ['John SMITH', 'Mickael JOHNSON']
        values = [doc.full_name() for doc in UserDocument.find().sort('_id').limit(2).skip(1)]

        self.assertEqual(values, expected_values)


class DocumentConfigurationTestCase(unittest.TestCase):

    def setUp(self):
        ConnectionManager.configure()

    def tearDown(self):
        UserDocument.col.remove()
        for attr in ('collection_name', 'config_name'):
            setattr(UserDocument, attr, None)

    def test_default_configuration(self):
        self.assertEqual(UserDocument.con.HOST, 'localhost')
        self.assertEqual(UserDocument.con.PORT, 27017)
        self.assertEqual(UserDocument.db.name, 'test')
        self.assertEqual(UserDocument.col.name, 'userdocument')

    def test_configuration_use_collection_attr(self):
        collection_name = 'my_collection'
        UserDocument.collection_name = collection_name

        self.assertEqual(UserDocument.col.name, collection_name)

    def test_configuration_use_config_name(self):
        config_name = 'my_custom_config_name'
        UserDocument.config_name = config_name

        db_name = 'my_custom_db'

        ConnectionManager.configure({config_name: {'db': db_name}})

        self.assertEqual(UserDocument.db.name, db_name)
        self.assertEqual(UserDocument.col.database.name, db_name)

    def test_configuration_use_col_from_config(self):
        config_name = 'my_custom_config_name'
        UserDocument.config_name = config_name

        col_name = 'my_custom_col'
        ConnectionManager.configure({config_name: {'col': col_name}})

        self.assertEqual(UserDocument.col.name, col_name)

class DocumentIndexesTestCase(unittest.TestCase):

    def setUp(self):
        ConnectionManager.configure()

    def tearDown(self):
        SimpleIndex.col.drop_indexes()
        SimpleIndex.col.remove()

    def test_simple_index(self):
        index = {'fields': ('something',)}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        self.assertEqual(mock_ensure_index.call_args_list, [Call('something', background=True)])

    def test_simple_descending_index(self):
        index = {'fields': ('-something',)}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        call_indexes = [('something', pymongo.DESCENDING)]
        self.assertEqual(mock_ensure_index.call_args_list, [Call(call_indexes, background=True)])

    def test_multiple_index(self):
        index = {'fields': ('something', 'something_else')}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        call_indexes = [('something', pymongo.ASCENDING),
                        ('something_else', pymongo.ASCENDING)]
        self.assertEqual(mock_ensure_index.call_args_list, [Call(call_indexes, background=True)])

    def test_mutliple_combined(self):
        index = {'fields': ('something', '-something_else')}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        call_indexes = [('something', pymongo.ASCENDING),
                        ('something_else', pymongo.DESCENDING)]
        self.assertEqual(mock_ensure_index.call_args_list, [Call(call_indexes, background=True)])

    def test_additionnals_args(self):
        index = {'fields': ('something',), 'unique': True, 'ttl': 3600 * 24}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        index.pop('fields')
        index['background'] = True
        self.assertEqual(
            mock_ensure_index.call_args_list,
            [Call('something', **index)])

    def test_default_background(self):
        index = {'fields': ('something',), 'unique': True, 'ttl': 3600 * 24}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        index.pop('fields')
        index['background'] = True
        self.assertEqual(
            mock_ensure_index.call_args_list,
            [Call('something', **index)])

    def test_override_background(self):
        index = {'fields': ('something',), 'unique': True, 'background': False}

        SimpleIndex.indexes = [index]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            SimpleIndex.generate_index()

        index.pop('fields')
        self.assertEqual(
            mock_ensure_index.call_args_list,
            [Call('something', **index)])

    def test_generate_index_return(self):
        index = {'fields': ('something',)}

        SimpleIndex.indexes = [index]

        def callable_ensure_index(*args, **kwargs):
            return args[0]

        with patch.object(SimpleIndex, 'col') as mock_col:
            mock_ensure_index = mock_col.ensure_index
            mock_ensure_index.side_effect = callable_ensure_index
            results = SimpleIndex.generate_index()

        self.assertEqual(results, ['something'])

    def test_functionnal_index(self):
        index = {'fields': ('something',), 'unique': True, 'ttl': 3600 * 24}

        SimpleIndex.indexes = [index]
        SimpleIndex.generate_index()

        s = SimpleIndex()
        s.something = 'something'
        s.save(safe=True)

        s2 = SimpleIndex()
        s2.something = 'something'
        self.assertRaises(DuplicateKeyError, s2.save, safe=True)

    def test_multiple_generate(self):
        SimpleIndex.generate_index()
        SimpleIndex.generate_index()

class DocumentPropertiesTestCase(unittest.TestCase):

    def test_properties_setter(self):
        class CustomException(Exception):
            pass

        class TestSetter(Document):

            @property
            def test(self):
                pass

            @test.setter
            def test(self, value):
                raise CustomException()

        t = TestSetter()
        def setter():
            t.test = 'something'
        self.assertRaises(CustomException, setter)

class DocumentDefaultsTestCase(unittest.TestCase):

    def setUp(self):
        ConnectionManager.configure()

    def tearDown(self):
        DefaultDocument.col.remove()

    def test_simple(self):

        defaults = {'a': 1, 'b': 2, 'c': 3}

        DefaultDocument.default_values = defaults

        default_doc = DefaultDocument()

        self.assertEqual(default_doc, defaults)

    def test_callable(self):
        callable_value = 'sample'

        def sample_callable():
            return callable_value

        defaults = {'a': 1, 'b': sample_callable, 'c': 3}

        DefaultDocument.default_values = defaults

        default_doc = DefaultDocument()

        self.assertEqual(default_doc, {'a': 1, 'b': callable_value, 'c': 3})

    def test_no_defaults(self):

        defaults = {'a': 1, 'b': 2, 'c': 3}

        DefaultDocument.default_values = defaults

        default_doc = DefaultDocument(use_defaults=False)

        self.assertEqual(default_doc, {})

    def test_find_one_no_defaults(self):

        defaults = {'a': 1, 'b': 2, 'c': 3}

        DefaultDocument.default_values = defaults

        doc_id = DefaultDocument.col.insert({})

        self.assertEqual(DefaultDocument.find_one(), {'_id': doc_id})

    def test_find_no_defaults(self):
        defaults = {'a': 1, 'b': 2, 'c': 3}

        DefaultDocument.default_values = defaults

        doc_id_1 = DefaultDocument.col.insert({})
        doc_id_2 = DefaultDocument.col.insert({})
        doc_id_3 = DefaultDocument.col.insert({})

        self.assertEquals(list(DefaultDocument.find().sort('_id')),
                          [{'_id': doc_id_1}, {'_id': doc_id_2},
                           {'_id': doc_id_3}])

    def test_mutable_default(self):
        defaults = {'a': []}
        DefaultDocument.default_values = defaults

        doc1 = DefaultDocument()
        self.assertEqual(doc1.a, [])
        doc1.a.append('b')
        self.assertEqual(doc1.a, ['b'])

        doc2 = DefaultDocument()
        self.assertEqual(doc2.a, [])


class NestedDocumentTestCase(unittest.TestCase):

    def tearDown(self):
        DefaultDocument.col.remove()

    def test_nested_document_find_one(self):
        doc = DefaultDocument()
        doc.nested = {'nested': 'nested'}
        doc.save()

        db_doc = DefaultDocument.find_one()
        self.assertEqual(doc, db_doc)
        self.assertTrue(isinstance(db_doc, DefaultDocument))

    def test_nested_document_find(self):
        doc = DefaultDocument()
        doc.nested = {'nested': 'nested'}
        doc.save()

        db_doc = DefaultDocument.find()[0]
        self.assertEqual(doc, db_doc)
        self.assertTrue(isinstance(db_doc, DefaultDocument))

class DocumentValidationTestCase(unittest.TestCase):

    def setUp(self):
        self.old_validate = ValidationDocument.validate

    def tearDown(self):
        ValidationDocument.validate = self.old_validate
        ValidationDocument.col.remove()

    def test_validation(self):
        def validate(self):
            pass

        ValidationDocument.validate = validate
        doc = ValidationDocument()
        doc.save(validate=True)

    def test_validation_default_fail(self):
        doc = ValidationDocument()

        self.assertRaises(NotImplementedError, doc.save, validate=True)

    def test_validation_fail(self):
        exception = Exception('My custom exception')

        def validate(self):
            raise exception

        ValidationDocument.validate = validate
        doc = ValidationDocument()
        self.assertRaises(Exception, doc.save, validate=True)

        try:
            doc.validate()
        except Exception as e:
            self.assertEqual(e, exception)

    def test_validation_change_forbidden(self):

        def validate(self):
            self.x = 1

        ValidationDocument.validate = validate
        doc = ValidationDocument({'x': 0})

        self.assertRaises(ValidationError, doc.save, validate=True)


    def test_validation_change_forbidden(self):

        def validate(self):
            del self.x

        ValidationDocument.validate = validate
        doc = ValidationDocument({'x': 0})

        self.assertRaises(ValidationError, doc.save, validate=True)

class DocumentPrivateProtected(unittest.TestCase):

    def setUp(self):
        self.doc = Document()

    def tearDown(self):
        Document.col.remove()

    def test_private_attr_id(self):
        """We must be able to set the _id of a document
        """
        oid = ObjectId()
        self.doc._id = oid
        self.doc.save()

        print "Self doc", self.doc

        self.assertEqual(Document.col.find_one()['_id'], oid)

    def test_private_attr_during_save(self):
        self.doc.a = 'a'
        self.doc._b = 'b'

        self.doc.save()

        self.assertEqual(Document.find_one(), {'_id': self.doc._id,
            'a': self.doc.a})

class DocumentCursorTestCase(unittest.TestCase):

    def tearDown(self):
        Document.col.remove()

    def test_count(self):
        Document.col.insert({})
        Document.col.insert({})

        self.assertEqual(Document.find().count(), 2)

    def test_advanced_cursor_methods(self):
        for i in range(10):
            Document.col.insert({'i': i})

        objects = Document.find().skip(2).limit(5)

        self.assertEqual(range(2, 7), [o['i'] for o in objects])


picomongo/utils.py
from connection_manager import ConnectionManager

#Proxy

def _class_name(cls):
    return cls.__name__.lower()

class CMProxy(object):

    def __init__(self, attr_name):
        self.attr_name = attr_name

    def __get__(self, instance, owner):
        config_name = owner.config_name

        config_name = config_name if config_name else _class_name(owner)
        return getattr(ConnectionManager.get_config(config_name),
                       self.attr_name)

class CollectionDescriptor(object):

    def __get__(self, instance, owner):
        class_name = _class_name(owner)
        config_name = owner.config_name
        config_name = config_name if config_name else class_name

        config_col = ConnectionManager.get_config(config_name).col
        if not config_col:
            collection_name = owner.collection_name
            document_name = collection_name if collection_name else class_name
            return ConnectionManager.get_config(config_name).db[document_name]

        return config_col


picomongo/document.py
from copy import copy, deepcopy

import pymongo

from pymongo.errors import InvalidOperation, OperationFailure
from pymongo.cursor import Cursor as PymongoCursor
from pymongo.read_preferences import ReadPreference

from exceptions import ValidationError
from utils import CMProxy, CollectionDescriptor

class DocumentCursor(PymongoCursor):

    def __init__(self, cursor, document_class, **kwargs):
        self.__dict__ = cursor.__dict__
        self._document = document_class
        self._kwargs = kwargs

    def __getitem__(self, index):
        return self._document(PymongoCursor.__getitem__(self, index),
                              **self._kwargs)

    def __getattr__(self, attr_name):
        return PymongoCursor.__getattribute__(self, attr_name)

    def next(self):
        return self._document(PymongoCursor.next(self), **self._kwargs)

class Document(dict):
    '''Base class for all documents.
    '''
    con = CMProxy('con')
    db = CMProxy('db')
    col = CollectionDescriptor()

    required_fields = []
    default_values = {}

    config_name = None
    collection_name = None

    def __init__(self, initial_values=None, use_defaults=True):
        init = {}
        if use_defaults:
            for key, value in self.default_values.items():
                init[key] = value() if hasattr(value, '__call__') else \
                    deepcopy(value)
        if initial_values is None:
            initial_values = {}

        if any(key.startswith('_') and key != "_id" for key in initial_values):
            raise ValueError("Initial values should not contains private fields")

        init.update(initial_values)

        super(Document, self).__init__(init)

    def save(self, validate=False, reload=False, **kwargs):
        '''Save document in db. Does not save attribute starting with '_'.
        '''
        if validate:
            local_copy = copy(self)
            self.__class__.validate(local_copy)
            if local_copy != self:
                err_msg = 'Changes and deletion are forbiden in validate method'
                raise ValidationError(err_msg)

        # TODO: Should picomongo manage db error
        self.col.save(self, **kwargs)

        if reload:
            self.reload()

    @classmethod
    def find_one(cls, *args, **kwargs):
        '''Get a single document from the database and return it as a Document.

        Any additionnal arguments will be passed to Collection.find_one
        '''
        the_one = cls.col.find_one(*args, **kwargs)
        if the_one:
            return cls(the_one, use_defaults=False)
        return the_one

    @classmethod
    def find(cls, *args, **kwargs):
        '''Query the database and returns results as Documents.

        Any additionnal arguments will be passed to Collection.find
        '''
        return DocumentCursor(cls.col.find(*args, **kwargs), cls,
                              use_defaults=False)

    @classmethod
    def generate_index(cls):
        '''Generate index in DB using Document.indexes

        Generate index use attribute named 'indexes' and accept theses format:
        * {'fields': ('something',)}: A single key index on the key 'something'
        * {'fields': ('-something',)}: A single descending key index on the key
            'something'
        * {'fields': ('something', '-something_else')}: A compound index on
            'something' ascending and 'something_else' descending

        Any additionnal items in 'indexes' will be passed as keyword arguments
            to ensure_index. For example:

        {'fields': ('something',), 'unique': True, 'ttl': 3600 * 24}
        '''

        ascending = lambda field: (field, pymongo.ASCENDING)
        # field[1:] remove '-' from field name
        descending = lambda field: (field[1:], pymongo.DESCENDING)

        results = []
        for index in cls.indexes:
            index = index.copy()

            fields = index.pop('fields')

            if len(fields) > 1:
                fields = [descending(field) if field.startswith('-')
                          else ascending(field) for field in fields]
            elif fields[0].startswith('-'):
                fields = [descending(fields[0])]
            else:
                fields = fields[0]

            if 'background' not in index:
                index['background'] = True
            results.append(cls.col.ensure_index(fields, **index))
        return results

    def reload(self):
        '''Reload current document from DB.

        Any additionnal arguments will be passed to Document.find_one
        Raise an InvalidOperation if current document is unsaved.
        '''
        if not self.get('_id'):
            raise InvalidOperation('You cannot reload an unsaved document.')

        doc = self.col.find_one({'_id': self._id}, read_preference=ReadPreference.PRIMARY)

        if not doc:
            raise OperationFailure('Document is no more present in DB.')

        self.clear()
        self.update(doc)

    def delete(self, *args, **kwargs):
        '''Remove current Document from database.

        Any additionnal arguments will be passed to Collection.remove
        Raise an InvalidOperation if current document is unsaved.
        '''
        if not '_id' in self:
            raise InvalidOperation('You cannot remove an unsaved document.')
        return self.col.remove({'_id': self._id}, *args, **kwargs)

    def validate(self):
        '''Override this method to add document validation.

        It takes no arguments.
        If it raises an exception, validation will be considered as failed.
        Otherwise validation will be considered as a success.
        '''
        msg = 'Validate method is not implemented for this document model.'
        raise NotImplementedError(msg)

    def __str__(self):
        return '%s(%s)' % (self.__class__.__name__, dict(self))

    def __repr__(self):
        return self.__str__()

    __getattr__ = dict.__getitem__

    def __setattr__(self, attr_name, value):
        white_list = set(('_id',)) # Set
        # Hack for properties setter
        if hasattr(getattr(self.__class__, attr_name, None), '__set__'):
            return object.__setattr__(self, attr_name, value)
        else:
            if attr_name.startswith('_') and attr_name not in white_list:
                return object.__setattr__(self, attr_name, value)
            else:
                return self.__setitem__(attr_name, value)

    def __delattr__(self, attr_name):
        if attr_name.startswith('_'):
            return object.__delattr__(self, attr_name)
        else:
            return self.__delitem__(attr_name)
